[NFC][Clang][Docs] Update Pointer Authentication documentation

ojhunt opened the pull request:
This updates the pointer authentication documentation to include a complete description of the existing functionaliy and behaviour, details of the more complex aspects of the semantics and security properties, and the Apple arm64e ABI design.

Co-authored-by: Ahmed Bougacha
Co-authored-by: Akira Hatanaka
Co-authored-by: John Mccall

ojhunt committed:
[NFC][Clang][Docs] Update Pointer Authentication documentation

This updates the pointer authentication documentation to include
a complete description of the existing functionaliy and behaviour,
details of the more complex aspects of the semantics and security
properties, and the Apple arm64e ABI design.

Co-authored-By: Ahmed Bougacha <ahmed@bougacha.org>
Co-authored-By: Akira Hatanaka <ahatanak@gmail.com>
Co-authored-By: John Mccall <rjmccall@apple.com>

asl commented on clang/docs/PointerAuthentication.rst:
```diff
@@ -163,14 +163,20 @@ a cryptographic signature, other implementations may be possible.  See
     data is simply a pepper added to the hash, not an encryption key, and so
     can be initialized using random data.
 
+ *** FIXME *** fpac?
```
Looks like it was mentioned?

ojhunt replied:
yup, I would _swear_ I stripped those out after I got that all written up

asl commented on clang/docs/PointerAuthentication.rst:
Do you want to say something that type discrimination for function pointers still need to treat different integer types as a single type?

asl reviewed (commented):
Looks great to me! Couple of minor things.

Tagging @atrosinenko @kbeyls @smithp35 for visibility

zwuis commented:
Nit: if you want GitHub to know that this is co-authored commit, you need to use the format `Co-authored-by: name <email@address.com>`. See <https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors>.

AaronBallman commented on clang/docs/PointerAuthentication.rst:
```diff
@@ -47,16 +47,16 @@ This document serves four purposes:
-  Objective-C, and Swift on arm64e, although these are not yet stable on any
-  target.
+- It documents the language ABIs currently used for C, C++, and Objective-C
+  on arm64e.
```
Any stability guarantees? outside of arm64e?

ojhunt replied:
These are our platform ABI so no significant changes are possible.

AaronBallman replied:
Probably worth spelling that out explicitly then.

ojhunt replied:
Clarified this is the stable abi for arm64e platform specifically.

AaronBallman commented on clang/docs/PointerAuthentication.rst:
```diff
@@ -255,33 +260,160 @@ signing schema breaks down even more simply:
-- The language provides the ``<ptrauth.h>`` intrinsic interface for manually
-  signing and authenticating pointers in code.  These can be used in
+- Finally, the language provides the ``<ptrauth.h>`` intrinsic interface for
+  manually signing and authenticating pointers in code.  These can be used in
```
Would one of those common circumstances be when loading a function pointer dynamically like via `dlsym` or `GetProcAddress`? If so, maybe we should call that out here?

AaronBallman replied:
How does the linker do that when these are dynamically loaded at runtime?

AaronBallman replied:
> 1. you would use the macros from `ptrauth.h` anytime you have a non-standard or complex signing schema - we have to use these macros in unwinding as the pc and stack pointers are counter signed. That's an idiom that is valuable in other contexts as well - for example, if you have ABI constraints that require a type be memcopyable you may want to adopt a custom schema where fields get countersigned so that you can ensure the overall struct is not compromised/intermingled without incurring the abi breakage the address discrimination would cause.

Ah, thank you! Do you think it's worth mentioning this in the docs?

> 2. GetProcAddress would be expected to return a correctly signed pointer. The user does not need to do anything themselves: the linker is trust root, and has to ensure that what it does vend are valid and correctly signed.

Ah, that should probably be explicitly documented. I was thinking you'd have the ability to do something yourself here, because I'm used to Win32 APIs like [`EncodePointer` and `DecodePointer`](https://learn.microsoft.com/en-us/previous-versions/bb432254(v=vs.85)) where you can manually protect pointers.

ojhunt replied:
I've worked out how to do this reasonably.

ojhunt replied:
Added a new section on dynamic loading

AaronBallman commented on clang/docs/PointerAuthentication.rst:
```diff
@@ -255,33 +260,160 @@ signing schema breaks down even more simply:
+
+For the most part, pointer authentication is an unobserved detail of the
+implementation of the programming language.  Any element of the language
+implementation that would perform an indirect branch to a pointer is implicitly
```
"indirect branch to a pointer" -- computed gotos (https://gcc.gnu.org/onlinedocs/gcc/Labels-as-Values.html)?

ojhunt replied:
computed goto, function pointers, virtual functions, etc

ojhunt replied:
@AaronBallman finally worked it out - computed goto protection is performed as part of the lowering in llvm, which has IR tests. I'll add a clang front end -> native code test on the assumption that it is possible for a clang error to result in lowering not recognizing the need for ptrauth protection.

AaronBallman commented on clang/docs/PointerAuthentication.rst:
```diff
@@ -255,33 +260,160 @@ signing schema breaks down even more simply:
+
+While this representation of nulls is the safest option for the general case,
+there are some situations in which a null pointer may have important semantic
+or security impact. For that purpose clang has the concept of a pointer
```
```suggestion
or security impact. For that purpose, Clang has the concept of a pointer
```

ojhunt replied:
(resolved locally if not pushed already)

kbeyls reviewed (commented):
Thank you for providing such well-written, detailed, and necessary documentation!

I left a few minor comments inline.

ojhunt commented on clang/docs/PointerAuthentication.rst:
There is currently no MS ABI support for ptrauth at all - if MS does want to adopt it they'll probably want to make choices here that actually make sense for their ABI, symbol naming, etc rather than ad hoc "abi" from people with only a passing familiarity with their ABI and ABI constraints :)

AaronBallman commented on clang/docs/PointerAuthentication.rst:
I suppose the compiler is responsible for signing and authenticating the core registers? Mostly, when I hear `setjmp`/`longjmp` I think exceptions because 1) that's the C form of exception handling, and 2) exceptions are sometimes implemented via those interfaces even for C++. Do we do the handling for the user regardless of which exception model they're using?

ojhunt replied:
the runtime knows the authentication schema of the registers, and re-signs them to a different schema as it transfers them into and out of the register buffers used for setjmp/longjmp, unwinding, etc.

ojhunt commented on clang/docs/PointerAuthentication.rst:
No - blocks need this as their various metadata fields are exposed ABI.

ojhunt replied:
oh sorry, there are two questions: the abi and the mangling. I believe "no" is the answer to both due to the way lambdas are exposed to the abi.

ojhunt commented on clang/docs/PointerAuthentication.rst:
```diff
@@ -327,6 +464,54 @@ a discriminator determined as follows:
+- A union type is both illegal and non-trivial to copy if it has a field whose
+  type is non-trivial or illegal to copy.
+
+- A union type is address-sensitive if it has a field whose type is
```
Lambdas don't have the ABI constraints of blocks, so mercifully there is not any ABI impact (because any compilation of a lambda is functionally a unique type and can change across compilations anyway)

ojhunt commented on clang/docs/PointerAuthentication.rst:
- [ ] add explanation

ojhunt commented on clang/docs/PointerAuthentication.rst:
It's talking about read only data in the program binary, I'll clarify that, and having two points for the same attack vector isn't super clear

- [ ] todo

ojhunt replied:
Updated to clarify this is about readonly data rather than the code section (the first bullet point in this section covers "attacker can directly modify/inject code"

ojhunt commented on clang/docs/PointerAuthentication.rst:
- [ ] todo

ojhunt commented on clang/docs/PointerAuthentication.rst:
nope, it's because I wrote an incoherent sentence :D

ojhunt replied:
Made this more coherent :D

ojhunt committed:
Address _some_ of the review feedback

Also update to __has_feature, though that is blocked on #153291

ojhunt committed:
More substantive doc updates

Adds more information (and documents arm64e abi) for dynamic symbol
resolution.

Clarifies remap of readonly sections

Fix incoherent sentence about the ptrauth_vtable_pointer attribute

ojhunt committed:
Merge remote-tracking branch 'origin/main' into users/ojhunt/ptrauth-docs

ojhunt committed:
Address feedback and small fixes

* Questions and clarifications from @kbeyls
* Fixing another example of Oliver breaking doc generation
* A couple of other small clarifications

ojhunt committed:
Update to prioritise __PTRAUTH__ where possible

AaronBallman reviewed (approved):
LGTM!

ojhunt committed:
Add a bit more of a note to clarify that data protection is not solely for code protection

ojhunt committed:
Merge branch 'main' into users/ojhunt/ptrauth-docs
